text/0119-dynamic-component-creation.md

```diff
@@ -0,0 +1,338 @@
+    }
+}
+```
+
+Renders:
+
+Before import completes:
+
+```html
+<div class="container"></div>
+```
+
+After import completes:
+
+```html
+<div class="container">
+    <lightning-concrete-component></lightning-concrete-component>
+</div>
+```
+
+## Motivation
+
+The main issue with the current implementation of the `lwc:dynamic` directive is that it does not respect the 1:1 mapping of the custom element name to constructor in the `CustomElementRegistry`.  This directly conflicts with the [custom elements specification](https://html.spec.whatwg.org/multipage/custom-elements.html#custom-element) and will make adopting the native `CustomElementRegistry` more difficult in the future.
+
+For context, the current design requires a component author define an arbitrary custom element name which will be used as the name for all constructors passed to the `lwc:dynamic` directive, creating a 1:N mapping.
+
+For example:
+
+```html
+<template>
+  <x-lazy lwc:dynamic={customCtor}></x-lazy>
+</template>
+```
+
+```javascript
+import { LightningElement } from "lwc";
+
+export default class DynamicCtor extends LightningElement {
+  customCtor;
+
+  connectedCallback() {
+    this.loadCtor();
+  }
+
+  async loadCtor() {
+    const { default: ctor } = await import("c/customConstructor");
+    this.customCtor = ctor;
+  }
+}
+```
+
+Will always render the same tag regardless of the constructor.
+
+```html
+  <x-lazy>
+    <!-- c-custom-constructor content -->
+  </x-lazy>
+```
+
+This proposal aims to overcome the issues with the current design by exposing the custom element name at compile time and using it as the name of the dynamic component that is created at runtime.
+
+## Prior art
+
+Several other frameworks have the concept of dynamically rendering components:
+
+- [Svelte: `<svelte:component this={expression}>`](https://svelte.dev/docs#template-syntax-svelte-component)
+- [Vue: `<component :is="...">`](https://vuejs.org/guide/essentials/component-basics.html#dynamic-components)
+- [Angular: viewContainerRef](https://angular.io/guide/dynamic-component-loader)
+- [Lit: static expressions](https://lit.dev/docs/templates/expressions/#static-expressions)
+- [Aura: $A.createComponent](https://developer.salesforce.com/docs/atlas.en-us.lightning.meta/lightning/js_cb_dynamic_cmp_async.htm)
+
+Frameworks that use JSX such as [React](https://reactjs.org/docs/rendering-elements.html#rendering-an-element-into-the-dom) and [Stencil](https://stenciljs.com/docs/templating-jsx) can dynamically render components directly into the markup using expressions.
+
+The goal of managing dynamic components in LWC is to follow the same component lifecycle as all other LWC custom elements.  Having said that, taking a similar approach as Svelte and Vue seems the most natural and intuitive path forward.
+
+## Detailed design
+
+A new property, `lightningElementName` on `LightningElementConstructor` is introduced to store the name of the custom element.
```

Comment:
In our call, we agreed that this must not be observable to user-land... and registerComponent is a good way to achieve that.

Reply:
I fleshed out the _Storing and retrieving a custom element's name_ section to call this out.